plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -206,20 +211,44 @@ type CekEvaluationExceptionCarrying term fun =
     EvaluationException CekUserError (MachineError fun term) term
 
 -- See Note [Being generic over @term@ in 'CekM'].
--- | A generalized version of 'CekM' carrying a @term@.
--- 'State' is inside the 'ExceptT', so we can get it back in case of error.
+-- | The monad the CEK machine runs in.
 -- The 'cost' parameter is for keeping track of costing in the 'StateT' monad.
-type CekCarryingM cost term uni fun s =
+type CekM cost uni fun s =
     ReaderT (CekEnv cost uni fun s)
-        (ExceptT (CekEvaluationExceptionCarrying term fun)
             (StateT cost
-                (ST s)))
+                (ST s))
 
 -- | The CEK machine-specific 'EvaluationException'.
 type CekEvaluationException uni fun = CekEvaluationExceptionCarrying (Term Name uni fun ()) fun
 
--- | The monad the CEK machine runs in.
-type CekM cost uni fun s = CekCarryingM cost (Term Name uni fun ()) uni fun s
+-- | The set of constraints we need to be able to print things in universes, which we need in order to throw exceptions.
+type PrettyUni uni fun = (GShow uni, Closed uni, Pretty fun, Typeable uni, Typeable fun, Everywhere uni PrettyConst)
+
+{- Note [Throwing exceptions in ST]
+This note represents MPJ's best understanding right now, might be wrong.
+
+We use a moderately evil trick to throw exceptions in ST, but unlike the evil trick for catching them, it's hidden.
+
+The evil is that the 'MonadThrow' instance for 'ST' uses 'unsafeIOToST . throwIO'! Sneaky! The author has marked it
+"Trustworthy"", no less. However, I believe this to be safe for basically the same reasons as our trick to catch
+exceptions is safe, see Note [Catching exceptions in ST]
```

Comment:
At that point, should we just use `IO` instead of pretending to be pure? I'm kinda tired of threading that `s` type variable everywhere.

Reply:
Hmm, I think it's still nice to have the non-`ST` stuff very closely guarded. It's still _mostly_ `ST`!

Reply:
@michaelpj Did you come up with this ST exception trick or can you provide a reference to it? I just want to learn more about it.

Reply:
I came up with it. There are some examples of people writing the generic version of this:
```
tryST :: (Exception e) => ST s a -> ST s (Either e a)
tryST act = unsafeIOToSt $ try $ unsafeSTToIO act
```
Which I think *is* unsafe because it's generic: this is not safe for all arguments you might pass in. However we're using it on a known ST action and catching a known exception type, which I think makes it okay.